backend/src/main/java/com/empire/World.java

```diff
@@ -2064,6 +2047,66 @@ void eatFood() {
 		}
 
 		void growPopulations() {
+			// Emigration
+			class Emigration {
+				final Region source;
+				final Region sink;
+				final double population;
+				public Emigration(Region source, Region sink, double population) {
+					this.source = source;
+					this.sink = sink;
+					this.population = population;
+				}
+			}
+			List<Emigration> emigrations = new ArrayList<>();
+			for (Region r : regions) {
+				double eligibleToEmigrate = r.population * r.unrestPopular * Constants.emigrationFactor;
+				if (starvingRegions.contains(r)) eligibleToEmigrate *= (1 + Constants.emigrationStarvationMod);
+				eligibleToEmigrate = Math.min(eligibleToEmigrate, r.population - 1); // Emigrating the last person causes a divide by zero error.
+				List<Region> destinations = new ArrayList<Region>();
+				for (Region n : r.getNeighbors(World.this)) {
+					if (n.type == Region.Type.LAND) destinations.add(n);
+					else for (Region nn : n.getNeighbors(World.this)) {
+						if (nn != r && n.type == Region.Type.LAND) destinations.add(nn);
+					}
+				}
+				destinations.removeIf(d -> d.unrestPopular >= r.unrestPopular);
+				destinations.removeIf(d -> starvingRegions.contains(d));
+				destinations.removeIf(d -> !d.getKingdom().equals(r.getKingdom()) && (patrolledRegions.contains(r) || getNation(d.getKingdom()).getRelationship(r.getKingdom()).refugees == Relationship.Refugees.REFUSE));
+				if (destinations.isEmpty() || eligibleToEmigrate < 1) continue;
+				double totalWeight = 0;
+				for (Region d : destinations) totalWeight += d.calcImmigrationWeight();
+				for (Region d : destinations) {
+					emigrations.add(new Emigration(r, d, d.calcImmigrationWeight() / totalWeight * eligibleToEmigrate));
+				}
+			}
+			Map<Region, Double> netMotion = new HashMap<>();
+			for (Emigration e : emigrations) {
```

Comment:
I would consider moving most of the inner part of this loop to the Emigration class as a sort of apply() method for doing all the population updating, and keep the netMotion-related calculation outside as is.

Reply:
Done.